### ISSUE TITLE
Could we have enhanced Comparators?

### ISSUE THREAD

**Issue by DanielGronau** [vavr-io/vavr#2143]
Javas comparator interface is very anaemic, and even worse has this unfortunate logic based on int return types. Could we have an interface extending `Comparator` with the following properties:

- There is an abstract method which returns an enum (with values like`LT`, `GT`, `EQ`) for a comparison, not an `int`
- there is a default implementation of the `compare` method delegating to the new comparison method
- it is easy to wrap existing `Comparators` to enhance them
- there are easy ways to chain enhanced comparators
- there are convenience methods with readable names like `isGreater`, `isGreaterOrEqual` etc

If there is agreement that this would be useful, and about the API, I could provide an implementation.

**Comment by chb0github** [comment]
You have the wrong programming language in mind. Try [kotlin](http://kotlination.com/kotlin/kotlin-compare-objects-example)

It has [extension functions](https://kotlinlang.org/docs/reference/extensions.html)

**Comment by DanielGronau** [comment]
I did exactly that in Kotlin, and it works great, but for Java we have inheritance only. However, I think even this is still better than the cold naked Comparator - you can't even have a switch on a comparision result, which feels like stone age. 

I have an implementation in my own lib ( https://github.com/highj/highj/blob/master/src/main/java/org/highj/data/ord/Ord.java ), but it has some bells and whistles attached which are not necessary here.

**Comment by danieldietrich** [comment]
Hi Daniel,

thank you for your suggestion! I've thoroughly thought about it. I see the benefit but think it does not fit well into Vavr.

Ord/Ordering is syntactic sugar for an existing concept. Beside the syntax aspect, the main design of Comparator is already okay, it is a functional interface and therefor allows pure implementations.

What is really missing in Java isn't syntactic sugar for the Comparator, it is a Comparable interface that behaves like Iterable.

* Iterable gives you an Iterator
* Comparable does *not* give you a Comparator

This makes it hard to create type hierarchies (like collections) where types are marked to have an underlying Comparator. In Vavr we added [Ordered](https://github.com/vavr-io/vavr/blob/988d252fcc09dc34290eaa6f2a713dc1de8514b1/vavr/src/main/java/io/vavr/collection/Ordered.java) for that purpose. However, in Vavr 1.0.0 it will be internally visible only because I want to expose as few types as possible to the outside.

My vision is to remove things from Vavr, step by step while Java evolves over time. The modularization of the library goes hand-in-hand with this strategy.

The most important principle is:

> We don't replace existing concepts (we make compromises) but we add missing features (while trying to get feature complete).

However, adding missing features must be differentiated:

* With **Match** and **For** we added pattern matching and for comprehensions. I think that was a mistake. These features could be handled efficiently by the compiler. We had to add them as runtime features, which has a significant performance impact. Also the syntax is not as good at it could be with native language support. If I could do it again, I would leave them away. With the same argument, I do not want a **Trampoline** for tailrec simulation, etc.
* **Match** and **For** are a good example for features that _might_ eventually come to Java, at least we will see pattern matching in not too far distance. Adding new features to Vavr increases the risk of diverging from the original language. Match is a good example. I've learned from the past and won't do that anymore.

There are three other examples where we replaced existing Java concepts:

* **Option** - Java has Optional _period_. We know that it is not designed to be a sum type and we have to live with it. If I needed to design a public API, I would use Optional whenever possible. I would use Vavr's Option locally because of its semantics and rich API, which plays well together with other Vavr types. Also Option is the only solution for serializable data objects with optional properties. So it is ok to diverge from native Java here by introducing the Option type, because it adds missing features that cannot accomplished in another way.
* **Iterator** - our Iterator isn't persistent, it is stateful. So it can be seen as replacement for java.util.Iterator. However, we need it under the hood for an efficient implementation of our collections. In fact it is very similar to java.util.stream.Stream - but more powerful in the sense of feature-richness.
* **Future** - when looking at CompletableFuture, I see a monolith that is a concept on its own. Some say, it is a monad but no one cared about abstracting over it by deeply integrating general concepts into the language - for example by supporting sum types. Sadly this concept will be missing when we will see native pattern matching. Compared to CompletableFuture, our Future can be used with almost the same syntax that other sum-types offer us.

It is a different story regarding our **functions** and **collections**. They do not replace existing Java types. Java does not have functions of arity > 2 or checked functions. Also Java does not have persistent collections.

You see, I want to talk as much Java as possible. By extending Comparator, it would integrate well but trading one or more new type(s) for syntactic sugar is too costly. My goal is to remove things for the upcoming 1.0.0 (or at least make them invisible to the outside).

I hope you understand that I think in this direction.

Thx

\- Daniel

PS: I think we should close the issue
